### PR TITLE
feat(gcp): Unify Project Discovery: Expose new interface for Org, Folder and Project filtering

### PR DESCRIPTION
**Pull request by bbernays**
#### Summary
closes: #11165 

The various existing Folder,Org and Projects filter/ id lists could not be used in combination to be able to express complex syncing strategies. 

This PR introduces a unified project discovery mechanism that enables users to simultaneously include and exclude based on filter/search requests as well hard coding IDs. This works for Organizations, Folders and Projects. Prior to this PR users could only express the resources they wanted to sync as long as they could fit the query in a single include filter statement.

```yaml
kind: source
spec:
  # Source spec section
  name: "gcp"
  registry: "grpc"
  path: localhost:7777
  tables: ["gcp_storage_buckets"]
  destinations: ["postgresql"]
  spec:
    hierarchy_discovery:
      organizations:
        # A hardcoded list of organizations to be synced
        include_list_id: 
        # A hardcoded list of organizations to be skipped during syncing
        exclude_list_id:
        # A list of queries where the results will be used as the list of organizations to be synced
        include_filter:
      folders:
        # A hardcoded list of folders to be synced
        include_list_id: 
        # A hardcoded list of folders to be skipped during syncing
        exclude_list_id:
        # A list of queries where the results will be used as the list of folders to be synced
        include_filter:
      projects:
        # A hardcoded list of projects to be synced
        include_list_id: 
        # A hardcoded list of projects to be skipped during syncing
        exclude_list_id:
        # A list of queries where the results will be used as the list of projects to be synced
        include_filter:
```

In a future PR I would want to remove the following attributes from the spec:
```
	ProjectIDs            []string `json:"project_ids"`
	FolderFilter          string   `json:"folder_filter"`
	FolderIDs             []string `json:"folder_ids"`
	FolderRecursionDepth  *int     `json:"folder_recursion_depth"`
	ProjectFilter         string   `json:"project_filter"`
	OrganizationIDs       []string `json:"organization_ids"`
	OrganizationFilter    string   `json:"organization_filter"`
```

### CODE REVIEW

File: plugins/source/gcp/client/spec.go [hunk 1]
```diff
@@ -4,6 +4,7 @@ package client
 type Spec struct {
 	ProjectIDs            []string `json:"project_ids"`
 	ServiceAccountKeyJSON string   `json:"service_account_key_json"`
+	FolderFilter          string   `json:"folder_filter"`
```
Comment (bbernays) [inline plugins/source/gcp/client/spec.go:7]: Need to add docs for this

File: plugins/source/gcp/client/spec.go [hunk 2]
```diff
@@ -4,6 +4,7 @@ package client
...
 	FolderIDs             []string `json:"folder_ids"`
 	FolderRecursionDepth  *int     `json:"folder_recursion_depth"`
```
Comment (bbernays) [inline plugins/source/gcp/client/spec.go:9]: In the future I think we should deprecate this. No need to use the recursion if we can just use a simple search query

**Review by erezrokah — APPROVED** [review]

**Comment by bbernays** [comment]
This PR doesn't actually solve the issue. This only enables users to use a different discovery method

**Review by erezrokah — COMMENTED** [review]
Thanks for the PR @bbernays, does having a single filter for each type of resource won't cover all uses cases?

For example if we have only a `projects_filter` users can:
```
include ids -> id:project-1 AND id:project-2
exclude ids -> NOT id:project-3
include + exclude names -> name:* AND NOT name:test*
```

**Comment by bbernays** [comment]
> Thanks for the PR @bbernays, does having a single filter for each type of resource will cover all uses cases?
> 
> For example if we have only a `projects_filter` users can:
> 
> ```
> include ids -> id:project-1 AND id:project-2
> exclude ids -> NOT id:project-3
> include + exclude names -> name:* AND NOT name:test*
> ```

There is nothing that stops users from doing that as it is in this implementation, but I thought that it would get very messy for users to have a single super long query rather than a list of queries... The goal of this was to keep the functionality it as explicit as possible

**Comment by erezrokah** [comment]
> There is nothing that stops users from doing that as it is in this implementation, but I thought that it would get very messy for users to have a single super long query rather than a list of queries... The goal of this was to keep the functionality it as explicit as possible

I'm not sure we should have multiple ways to do the same thing in this case. Maybe we could start only with the filter and add the more explicit methods later if needed? I'm worried that:
1. We're building an abstraction over GCP filter expressions
2. It's harder to communicate how our custom logic works instead of pointing to GCP docs and say we're passing through the expression directly so the logics matches `gcloud * list --filter`

We can add some examples in our docs on how to include/exclude project IDs, etc.

**Comment by bbernays** [comment]
I have removed the `exclude_filter` option, but left the `include_filter` to be explicit that the results will be included

**Comment by erezrokah** [comment]
> We already know that users use the hard-coded lists now why would we want to take away that user experience?

We are also adding exclude list which is more logic that implements on our end what the filter does on the backend.
I don't mind keeping the existing include list, but it does mean we'll be doing more API requests than needed to get the same data (I would even translate that list to a filter based query instead of iterating over the ids).

**Review by erezrokah — COMMENTED** [review]
Added a few comments but could we also maybe trim the API requests to a single filter?
For example:

1. User configures only filter -> use that directly
2. User configures only include list of IDs -> transform to an `OR` filter on all IDs
3. User configures both -> create a new filter that does an `OR` between the configured one an the ID based `OR` filter from `2.`

Am I missing a use case with the above approach?

File: plugins/source/gcp/client/spec.go [hunk 3]
```diff
@@ -1,9 +1,37 @@
 package client
 
+import "fmt"
+
+type ResourceDiscovery struct {
+	IncludeListId []string `json:"include_list_id"`
+	ExcludeListId []string `json:"exclude_list_id"`
+	IncludeFilter []string `json:"include_filter"`
```
[the remark this answers is not in the document]
Comment (bbernays) [reply plugins/source/gcp/client/spec.go:8]: Keeping it as an array of strings gives users more flexibility without a significant impact on usability... If we started with a string we would have to break backwards compatibility to add support...

**Comment by bbernays** [comment]
> > We already know that users use the hard-coded lists now why would we want to take away that user experience?
> 
> We are also adding exclude list which is more logic that implements on our end what the filter does on the backend. I don't mind keeping the existing include list, but it does mean we'll be doing more API requests than needed to get the same data (I would even translate that list to a filter based query instead of iterating over the ids).



> Added a few comments but could we also maybe trim the API requests to a single filter? For example:
> 
> 1. User configures only filter -> use that directly
> 2. User configures only include list of IDs -> transform to an `OR` filter on all IDs
> 3. User configures both -> create a new filter that does an `OR` between the configured one an the ID based `OR` filter from `2.`
> 
> Am I missing a use case with the above approach?

I really like that idea and I think this is a super interesting + powerful approach! I sort of want to do it in a follow up PR once we are aligned on the user facing pieces though

File: plugins/source/gcp/client/spec.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (erezrokah) [reply plugins/source/gcp/client/spec.go:8]: > Keeping it as an array of strings gives users more flexibility without a significant impact on usability... If we started with a string we would have to break backwards compatibility to add support...

What kind of flexibility are we getting? It seems we're implementing the `OR` functionality that already exists in the current filter syntax

**Comment by yevgenypats** [comment]
What is the status on that one? did we decide we keep the complexity at bay at the end and not introduce more options?

**Review by yevgenypats — CHANGES_REQUESTED** [review]
How do we intend to test this?

**Comment by erezrokah** [comment]
> What is the status on that one?

I'll spend some time today looking at the different built-in filtering logic GCP provides. Maybe we can assume some best practices on folders/projects/organizations structure to simplify filtering logic

**Comment by yevgenypats** [comment]
> > What is the status on that one?
> 
> I'll spend some time today looking at the different built-in filtering logic GCP provides. Maybe we can assume some best practices on folders/projects/organizations structure to simplify filtering logic

I think for such a complex feature which ads more complexity (and going back that I think we should simplify the current config even more and remove the org option and not add more options) we should wait for more users to ask for it and not rush it especially as there is a work-around.

**Comment by erezrokah** [comment]
Going to close this as stale, we still have the code as reference

**Pull request closed without merging**